src/reducers/scrollStateReducer.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright Schrodinger, LLC
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule scrollStateReducer
+ */
+
+'use strict';
+
+import IntegerBufferSet from 'IntegerBufferSet';
+import ActionTypes from 'ActionTypes'
+import isNil from 'lodash.isnil';
+import {
+  scrollBy,
+  scrollEnd,
+  scrollStart,
+  scrollTo,
+  updateRowCount,
+  updateRowHeights,
+  updateViewHeight,
+} from 'scrollStateHelper';
+
+const DEFAULT_STATE = {
+  rowsCount: 0,
+  rowHeightGetter: () => 0,
+  storedHeights: [],
+  viewportHeight: 0,
+  rowOffsets: null, //PrefixIntervalTree
+  scrollContentHeight: 0,
+  rowHeights: {},
+  rows: [], //rowsToRender
+  bufferSet: new IntegerBufferSet(),
+    
+  viewportRowsBegin: 0,
+  viewportRowsEnd: 0,
+  firstRowIndex: 0,
+  firstRowOffset: 0,
+  scrollY: 0,
+  scrollContentHeight: 0,
+  maxScrollY: 0
+};
+
+function scrollStateReducer(state = DEFAULT_STATE, action) {
+  switch (action.type) {
+    case ActionTypes.INITIALIZE:
+      let { props } = action;
+
+      state = updateRowCount(state, props);
+      state = updateRowHeights(state, props);
+      state = updateViewHeight(state, props);
```

Comment:
I also want to try and break up the helpers a bit more, so I created 
updateRowCount, updateRowHeights and updateViewHeight which do more piecemeal work.

When the props update, you'll only call the one's associated to your code. 
There is still a bit of code being re-run that we can optimize closer to release.